puma/hlplots/results.py

```diff
@@ -497,15 +501,116 @@ def plot_var_perf(  # pylint: disable=too-many-locals
+        plot_suffix = f"_{suffix}" if suffix else ""
+        plot_sig_eff.savefig(self.get_filename(plot_details + plot_base, plot_suffix))
+        for i, background in enumerate(self.backgrounds):
+            plot_bkg[i].draw()
+            plot_details = f"{background}_rej_vs_{x_var}_"
+            plot_bkg[i].savefig(
+                self.get_filename(plot_details + plot_base, plot_suffix)
+            )
+
+    def plot_flat_rej_var_perf(
+        self,
+        fixed_rejections: dict[Flavour, float],
+        suffix: str = None,
+        xlabel: str = r"$p_{T}$ [GeV]",
+        x_var: str = "pt",
+        h_line: float = None,
+        **kwargs,
+    ):
+        """Plot signal efficiency as a function of a variable, with a fixed enforce
+        background rejection for each bin
+
+
+        Parameters
+        ----------
+        fixed_rejections : dict[Flavour, float]
+            A dictionary of the fixed background rejections for each flavour, eg:
+            fixed_rejections = {cjets : 0.1, ujets : 0.01}
+        suffix : str, optional
+            suffix to add to output file name, by default None
+        xlabel : regexp, optional
+            _description_, by default "$p_{T}$ [GeV]"
+        x_var: str, optional
+            The x axis variable, used for providing details to the plot name,
+            default is 'pt'
+        h_line : float, optional
+            draws a horizonatal line in the signal efficiency plot
+        **kwargs : kwargs
+            key word arguments for `puma.VarVsEff`
+        """
+        assert all(
+            [b.name in fixed_rejections for b in self.backgrounds]
+        ), "Not all backgrounds have a fixed rejection"
+        plot_bkg = []
+        for background in self.backgrounds:
+            modified_second_tag = (
+                f"{self.atlas_second_tag}\nFixed {background.rej_str} ="
+                f" {fixed_rejections[background.name]} per bin"
+            )
+            plot_bkg.append(
+                VarVsEffPlot(
+                    mode="bkg_eff",
+                    ylabel=self.signal.eff_str,
+                    xlabel=xlabel,
+                    logy=False,
+                    atlas_first_tag=self.atlas_first_tag,
+                    atlas_second_tag=modified_second_tag,
+                    n_ratio_panels=1,
+                    y_scale=1.4,
+                )
+            )
+        for tagger in self.taggers.values():
+            if "disc_cut" in kwargs:
+                raise ValueError("disc_cut should not be set for this plot")
+            if "working_point" in kwargs:
+                raise ValueError("working_point should not be set for this plot")
+
+            discs = tagger.discriminant(self.signal)
+            is_signal = tagger.is_flav(self.signal)
+            for i, background in enumerate(self.backgrounds):
+                is_bkg = tagger.is_flav(background)
+                # We want x bins to all have the same background rejection, so we
```

Comment:
The approach I've gone for is a bit hacky, the current var_vs_eff code takes its mode, which defines what its putting on the y-axis. But applies WP cuts to the discriminant based on the signal. So, to plot signal efficiency on the y-axis, while keeping a flat background efficiency in each bin, I set the mode to background efficiency, and then pass the backgrounds as signal and vice-versa. This seems to give the correct result, and I couldn't think of a simpler way of doing this, but perhaps I'm just misunderstanding the main var_vs_eff code.

Reply:
That sounds fine to me, you could possibly just remove the specificity in the argument names, e.g. `x_var_sig` -> `xs`, `disc_sig` -> `disc`, and then mention in the docstring or in a comment that what these correspond to will depend on the mode

Reply:
So, i think I've decided to not modify these, and instead keep the majority of internals of class the same, but added a mode that allows for correct uncertainty calculations when running flat per bin rejections like this.

Reply:
> That sounds fine to me, you could possibly just remove the specificity in the argument names, e.g. `x_var_sig` -> `xs`, `disc_sig` -> `disc`, and then mention in the docstring or in a comment that what these correspond to will depend on the mode

I'm not sure if we can do this, currently we require 'x_var_sig', 'disc_sig', 'x_var_bkg', and 'disc_bkg', see [here](https://github.com/umami-hep/puma/blob/main/puma/var_vs_eff.py#L21-L24) which are all used in making the plots.
If we had just 'xs', 'disc', this would perhaps make things less clear?

Reply:
ah you are right, sorry for the noise